blx/storage_disk.go

```diff
@@ -0,0 +1,149 @@
+package blx
+
+import (
+	"encoding/gob"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// diskStorage stores the block in a file
+type diskStorage struct {
+	path string
+}
+
+const (
+	dataExt string = ".data"
+	metaExt        = ".meta"
+)
+
+// NewDiskStorage creates a new diskStorage struct with the given path
+// the files that will be generated from this struct are stored in the path
+func NewDiskStorage(path string) Storage {
+	os.MkdirAll(path, os.ModePerm)
+
+	return &diskStorage{
+		path: path,
+	}
+}
+
+// Store saves the block in two files one for the metadata and one for
+// the data. The convetion used is key.meta and key.data. Returns error if
+// the files cannot be created.
+func (d *diskStorage) Store(key string, block *Block) error {
+	dataFilePath := filepath.Join(d.path, block.Key+dataExt)
+	metaFilePath := filepath.Join(d.path, block.Key+metaExt)
+
+	dataFileFound := false
+	metaFileFound := false
+
+	// Check if both files exist otherwise overwrite them
+	if _, err := os.Stat(dataFilePath); err == nil {
+		dataFileFound = true
+	}
+
+	if _, err := os.Stat(metaFilePath); err == nil {
+		metaFileFound = true
+	}
+
+	if dataFileFound && metaFileFound {
+		return ErrExists
+	}
+
+	// Write the data in a file
+	err := ioutil.WriteFile(dataFilePath, block.Data, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Write the meta in a file
+	mf, err := os.Create(metaFilePath)
+	if err != nil {
+		return err
+	}
+
+	defer mf.Close()
+
+	// Create an encoder to store the meta map
+	enc := gob.NewEncoder(mf)
```

Comment:
Can we encode meta as JSON? GOB will break if we want to change the golang struct and JSON will be easier to check out what we have in the folder/s with jq.

Reply:
Didn't think of that, makes sense.